runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.serde
+
+import aws.smithy.kotlin.runtime.InternalApi
+
+/**
+ * This tag interface provides a mechanism to attach type-specific metadata to any field.
+ * See [aws.smithy.kotlin.runtime.serde.xml.XmlList] for an example implementation.
+ *
+ * For example, to specify that a list should be serialized in XML such that values are wrapped
+ * in a tag called "boo", pass an instance of XmlList to the FieldDescriptor of `XmlList(elementName="boo")`.
+ */
+@InternalApi
+public interface FieldTrait
+
+/**
+ * Indicates to deserializers to ignore field/key
+ *
+ * @param key The key to ignore in the payload
+ * @param regardlessOfInModel If true will ignore key even though the model indicates key should be there
+ *
+ */
+@InternalApi
+public data class IgnoreKey(public val key: String, public val regardlessOfInModel: Boolean = true) : FieldTrait
```

Comment:
This should be in `serde-json` as it only applies there right? (see `JsonFieldTraits.kt`)

Reply:
**Question**: Why do we need `regardlessOfInModel`, I would assume if we set `IgnoreKey` as a trait on the object descriptor it's going to be ignored regardless of whether it's modeled. This seems unnecessary to me but I'd like to understand what the motivation was.

Reply:
I was thinking of a case where we have a union with a member called `__type`. If that happens we would not deserialize that member, because we're ignoring that key regardless of what the model says. The default is true anyway so it behaves as expected unless set explicitly to false.

Reply:
I didn't put it in a JSON specific file because I think this might be useful in the future for other deserializers like the XML one. I don't know if we have a use case already but that was my reasoning behind it.

Reply:
> I was thinking of a case where we have a union with a member called __type. If that happens we would not deserialize that member, because we're ignoring that key regardless of what the model says

Shouldn't this just be a choice by codegen though? If we want it deserialized because it's in the model don't add the `IgnoreKey` trait.

Reply:
That makes sense but the issue this is trying to solve: https://github.com/smithy-lang/smithy/pull/1945
is that sometimes a unnecessary `__type` is serialized. What you're suggesting is to always ignore `__type`. But there could be a cases where `__type` could be in the model and we would be ignoring it because sometimes there's an extra one there unnecessarily

Reply:
> What you're suggesting is to always ignore __type. But there could be a cases where __type could be in the model and we would be ignoring it because sometimes there's an extra one there unnecessarily

I don't follow, if `__type` is in the model and we _want_ to deserialize it then just don't emit the `IgnoreKey` trait during codegen. What am I missing here?

Reply:
I guess that logic could live in codegen, now it's in the deserializer